libraries/chain/block_log.cpp

```diff
@@ -1475,4 +1508,238 @@ namespace eosio { namespace chain {
+   }
+
+   // When we already have an index file, it will be much faster to
+   // append missing indexes than build from the scratch
+   void trim_data::append_index(uint64_t num_indexes_existing) {
+      detail::reverse_iterator block_log_iter;
+
+      const uint32_t num_blocks = block_log_iter.open(block_file_name);
+      EOS_ASSERT(num_blocks > num_indexes_existing, block_log_exception, "Number of blocks in block file ${num_blocks} must be greater than blocks in index file ${num_indexes_existing}", ("num_blocks",num_blocks)("num_indexes_existing", num_indexes_existing));
+
+      uint32_t num_indexes_to_append = num_blocks - num_indexes_existing;
+
+      // Create a temporary file to store missing indexes
+      auto now = fc::time_point::now();
+      fc::path temp_file_name = index_file_name.generic_string().append("-").append( now );
+
+      // Construct missing indexes
+      detail::index_writer index(temp_file_name, num_indexes_to_append);
+      uint64_t position;
+      auto i = 0;
+      while (i < num_indexes_to_append && (position = block_log_iter.previous()) != block_log::npos) {
+         index.write(position);
+         ++i;
+      }
+      index.complete();
+
+      // Append missing index file to original index file
+      std::ofstream index_stream;
+      index_stream.open(index_file_name.generic_string().c_str(), LOG_WRITE);
+      std::ifstream temp_stream;
+      temp_stream.open(temp_file_name.generic_string().c_str(), LOG_READ);
+
+      EOS_ASSERT(index_stream.is_open(), block_log_exception, "index_file not open");
+      EOS_ASSERT(temp_stream.is_open(), block_log_exception, "temp_stream not open");
+
+      index_stream.seekp(0, std::ios_base::end);
+      index_stream << temp_stream.rdbuf();
+
+      index_stream.flush();
+
+      // Restore original index file handle
+      // Close original index file handle
+      fclose(ind_in);
+      ind_in = FC_FOPEN(index_file_name.generic_string().c_str(), "rb");
+
+      // Temp index file no longer needed
+      fc::remove(temp_file_name);
+   }
+
+   // Head block is corrupt. Find the last good block in block file,
+   // trim the bad portion, and make block file and index file 
+   // agree with each other
+   void trim_data::trim_to_last_good_block() {
+      // Get length of blocks.index (gives number of blocks)
+      const auto status = fseek(ind_in, 0, SEEK_END);
+      EOS_ASSERT( status == 0, block_log_exception, "cannot seek to ${file} end", ("file", index_file_name.string()) );
+      const uint64_t file_end = ftell(ind_in);
+      auto num_blocks = file_end/sizeof(uint64_t);
+
+      uint64_t block_pos = 0;
+      uint64_t block_number = 0;
+      bool good_block_found = false;
+
+      // Starting from last index in index file, look at its corresponding
+      // block in log file.
+      for (auto  i = 1; i <= num_blocks; ++i) {
```

Comment:
The type of ```i``` will be deduced as an ```int```.  So you are comparing mixed signed values (```num_blocks``` is unsigned).

Reply:
Agreed. This makes code cleaner.